src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+                    bool acquired;
+
+                    while (!(acquired = TryAcquireLockRecord(name, maxLifetime, out @lock)) && waitedTime < timeout) {
+                        Task.Delay(timeout).ContinueWith(t => {
+                            waitedTime += waitTime;
+                        }).Wait();
+                    }
+
+                    if (acquired) {
+                        Logger.Debug("Successfully acquired a lock named {0}.", name);
+                        return true;
+                    }
+                }
+                catch (Exception ex) {
+                    Logger.Error(ex, "Error while trying to acquire a lock named {0}.", name);
+                    throw;
+                }
+
+                Logger.Debug("Could not acquire a lock named {0}.", name);
+                return false;
+            }
+        }
+
+        public IDistributedLock AcquireLock(string name, TimeSpan maxLifetime, TimeSpan timeout) {
+            IDistributedLock lockResult;
+            return TryAcquireLock(name, maxLifetime, timeout, out lockResult) ? lockResult : null;
+        }
+
+        public void DisposeLock(IDistributedLock @lock) {
+            var childLifetimeScope = CreateChildLifetimeScope(@lock.Name);
+
+            try {
+                var repository = childLifetimeScope.Resolve<IRepository<LockRecord>>();
+                var transactionManager = childLifetimeScope.Resolve<ITransactionManager>();
+                transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+                var record = repository.Get(@lock.Id);
+
+                if (record != null) {
+                    if (record.ReferenceCount > 0)
+                        record.ReferenceCount--;
+                }
+            }
+            catch (Exception ex) {
+                if (ex.IsFatal()) throw;
+                Logger.Error(ex, "An non-fatal error occurred while trying to dispose a distributed lock with name {0} and ID {1}.", @lock.Name, @lock.Id);
+            }
+            finally {
+                childLifetimeScope.Dispose();
+            }
+        }
+
+        protected virtual bool TryAcquireLockRecord(string name, TimeSpan maxLifetime, out IDistributedLock @lock) {
+            @lock = null;
+            var childLifetimeScope = CreateChildLifetimeScope(name);
+
+            try {
+                var transactionManager = childLifetimeScope.Resolve<ITransactionManager>();
+                transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+
+                // This way we can create a nested transaction scope instead of having the unwanted effect
+                // of manipulating the transaction of the caller.
+                var repository = childLifetimeScope.Resolve<IRepository<LockRecord>>();
+
+                // Find an existing, active lock, if any.
+                var record = repository.Table.FirstOrDefault(x => x.Name == name && (x.ValidUntilUtc >= _clock.UtcNow || x.ReferenceCount > 0));
```

Comment:
Yes good catch. My original definition of an active (unavailable) lock was as follows:
- One that has not yet expired *or* one whose count is greater than 0.

However, it should of course be:
- One that has not yet expired *and* whose count is greater than 0.

I'll add a unit test to formalize this rule.